File: src/go-server/runtime_os.go
package main

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
)

type RuntimeOperatingSystem struct {
	Name   string
	Source string
}

func readRuntimeOperatingSystem(paths ...string) (RuntimeOperatingSystem, error) {
	for _, path := range paths {
		contents, err := os.ReadFile(path)
		if errors.Is(err, fs.ErrNotExist) {
			continue
		}
		if err != nil {
			return RuntimeOperatingSystem{}, fmt.Errorf("read runtime OS evidence %s: %w", path, err)
		}
		return RuntimeOperatingSystem{string(contents), path}, nil
	}
	return RuntimeOperatingSystem{"unexposed-by-runtime", "unexposed-by-runtime"}, nil
}

File: src/rust-server/src/runtime_os.rs
use std::{fs, io, path::Path};

pub struct RuntimeOperatingSystem {
    pub name: String,
    pub source: String,
}

pub fn read(paths: &[&Path]) -> Result<RuntimeOperatingSystem, String> {
    for path in paths {
        let contents = match fs::read_to_string(path) {
            Ok(contents) => contents,
            Err(error) if error.kind() == io::ErrorKind::NotFound => continue,
            Err(error) => return Err(format!("read runtime OS evidence {}: {error}", path.display())),
        };
        return Ok(RuntimeOperatingSystem {
            name: contents,
            source: path.to_string_lossy().into_owned(),
        });
    }
    Ok(RuntimeOperatingSystem {
        name: "unexposed-by-runtime".into(),
        source: "unexposed-by-runtime".into(),
    })
}
